airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py

```diff
@@ -193,29 +210,50 @@ def generate_streams(
                 )
                 continue
 
-            streams.append(self._wrap_for_concurrency(config, stream, state_manager))
-        streams.append(self._wrap_for_concurrency(config, Describe(sf_api=sf_object, catalog=self.catalog), state_manager))
+            converter = stream.state_converter
+
+            cursor = self.initialize_cursor(
+                stream,
+                state_manager,
+                converter,
+                self._get_slice_boundary_fields(stream, state_manager),
+                start,
+                converter.get_end_provider(),
+                lookback_window=lookback_window,
+                slice_range=slice_range,
+            )
+
+            streams.append(self._convert_to_concurrent_stream(logger, stream, cursor=cursor))
+
+        streams.append(self._convert_to_concurrent_stream(logger, describe_stream, cursor=describe_stream_cursor))
         return streams
 
-    def _wrap_for_concurrency(self, config, stream, state_manager):
-        stream_slicer_cursor = None
-        if stream.cursor_field:
-            stream_slicer_cursor = self._create_stream_slicer_cursor(config, state_manager, stream)
-            if hasattr(stream, "set_cursor"):
-                stream.set_cursor(stream_slicer_cursor)
-        if hasattr(stream, "parent") and hasattr(stream.parent, "set_cursor"):
-            stream_slicer_cursor = self._create_stream_slicer_cursor(config, state_manager, stream)
-            stream.parent.set_cursor(stream_slicer_cursor)
-
-        if not stream_slicer_cursor or self._get_sync_mode_from_catalog(stream) == SyncMode.full_refresh:
-            cursor = FinalStateCursor(
-                stream_name=stream.name, stream_namespace=stream.namespace, message_repository=self.message_repository
-            )
-            state = None
-        else:
-            cursor = stream_slicer_cursor
-            state = cursor.state
-        return StreamFacade.create_from_stream(stream, self, logger, state, cursor)
+    def initialize_cursor(
+        self,
+        stream: Stream,
+        state_manager: ConnectorStateManager,
+        converter: AbstractStreamStateConverter,
+        slice_boundary_fields: Optional[Tuple[str, str]],
+        start: Optional[CursorValueType],
+        end_provider: Optional[Callable[[], CursorValueType]] = None,
+        lookback_window: Optional[GapType] = None,
+        slice_range: Optional[GapType] = None,
+    ) -> Optional[ConcurrentCursor]:
+        cursor_field = stream.cursor_field
+
+        if cursor_field and not isinstance(cursor_field, str):
+            raise AssertionError(f"Nested cursor field are not supported hence type str is expected but got {cursor_field}.")
```

Comment:
should this be a `ValueError` instead?

Reply:
updated to validate it out of the box and raise a ValueError.